plumbing/format/packfile/parser.go

```diff
@@ -35,6 +35,33 @@ type Observer interface {
 	OnFooter(h plumbing.Hash) error
 }
 
+type StatusObserver struct {
+	statusChan plumbing.StatusChan
+	update     plumbing.StatusUpdate
+}
+
+func NewStatusObserver(sc plumbing.StatusChan) *StatusObserver {
+	return &StatusObserver{statusChan: sc}
+}
+func (so *StatusObserver) OnHeader(count uint32) error {
+	so.update.Stage = plumbing.StatusFetch
+	so.update.ObjectsTotal = int(count)
+	so.statusChan.SendUpdate(so.update)
+
+	return nil
+}
+func (so *StatusObserver) OnInflatedObjectHeader(t plumbing.ObjectType, objSize int64, pos int64) error {
+	so.statusChan.SendUpdate(so.update)
+	so.update.ObjectsDone++
```

Comment:
Also, are we guaranteed this will only be called from a single goroutine?  No need for a mutex?

Reply:
Observers are associated with a Parser and all the Parser internal data is accessed without any locking. So if there is concurrent access then the whole upstream parser abstraction is racy.